### PR TITLE
Add clickhouse2deps

### PR DESCRIPTION
**Pull request by r0mainK**
To be merged after #5 , this contains the model/logic/sql helpers, etc for extracting the dependencies from clickhouse to an ASDF model. There is no postprocessing or anything here, I just query the DB and save the results.

EDIT: Ill be adding tests when the core logic is approved.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,3 +1,7 @@
 sourced-ml-core==0.0.5
 numpy==1.16.2
 humanize==0.5.1
+clickhouse-driver==0.1.1
+modelforge==0.14.1
```
Comment (vmarkovtsev) [inline requirements.txt]: sourced-ml-core already depends on modelforge

File: setup.py [hunk 2]
```diff
@@ -40,9 +40,21 @@
         "sourced-ml-core>=0.0.5",
         "PyStemmer>=1.3,<2.0",
         "humanize>=0.5.0,<0.6",
+        "clickhouse-driver>=0.1.1,<1.0",
+        "modelforge>=0.14.1,<1.0",
```
Comment (vmarkovtsev) [inline setup.py]: Same here

File: setup.py [hunk 3]
```diff
@@ -40,9 +40,21 @@
...
+        "scipy>=1.0,<2.0",
+        "jinja2>=2.10.1<3.0",
     ],
     tests_require=["docker>=3.6.0,<4.0"],
-    package_data={"": ["LICENSE.md", "README.md"]},
+    package_data={
+        "": ["LICENSE.md", "README.md"],
+        "sourced": [
+            os.path.join(d, fp).replace("sourced/", "")
```
Comment (vmarkovtsev) [inline setup.py]: Now that we dropped 3.5, we can forget about these `os.path`-s and use much better `pathlib` API everywhere.

```python
from pathlib import Path

pkg_path = Path("sourced/ml/mining")

SourceFileLoader("...", pkg_path / "__init__.py")

(Path(__file__).parent / "README.md").open(encoding="utf-8")

pkg_path.rglob("*.json*") + pkg_path.rglob("*.jinja2")
```

File: sourced/ml/mining/__main__.py [hunk 4]
```diff
@@ -23,6 +27,53 @@ def add_parser(name, help_message):
             name, help=help_message, formatter_class=ArgumentDefaultsHelpFormatterNoNone
         )
 
+    # --------------------------------------------------------------------------------------------
+
+    clickhouse2deps_parser = add_parser(
+        "clickhouse2deps", "Extract dependencies from a Clickhouse instance."
```
Comment (vmarkovtsev) [inline sourced/ml/mining/__main__.py]:
```suggestion
        "clickhouse2deps", "Extract dependencies from a ClickHouse DB."
```

File: sourced/ml/mining/__main__.py [hunk 5]
```diff
@@ -23,6 +27,53 @@ def add_parser(name, help_message):
...
+    )
+    clickhouse2deps_parser.set_defaults(handler=clickhouse2deps)
+    clickhouse2deps_parser.add_argument(
+        "-o", "--output-path", help="Output path for the ASDF dependencies model."
```
Comment (vmarkovtsev) [inline sourced/ml/mining/__main__.py]:
```suggestion
        "-o", "--output-path", help="Output path to the resulting ASDF model with the extracted dependencies."
```

File: sourced/ml/mining/__main__.py [hunk 6]
```diff
@@ -23,6 +27,53 @@ def add_parser(name, help_message):
...
+    )
+    clickhouse2deps_parser.add_argument(
+        "-f",
+        "--force",
+        action="store_true",
+        help="Whether to overwrite existing ASDF model.",
```
Comment (vmarkovtsev) [inline sourced/ml/mining/__main__.py]:
```suggestion
        help="Value indicating whether to overwrite the existing ASDF model specified by -o/--output-path.",
```

File: sourced/ml/mining/__main__.py [hunk 7]
```diff
@@ -23,6 +27,53 @@ def add_parser(name, help_message):
...
+    clickhouse2deps_parser.add_argument(
+        "--password", default="", help="Password for the DB."
+    )
+    clickhouse2deps_parser.add_argument(
+        "--host", default="0.0.0.0", help="Host for the DB."
+    )
+    clickhouse2deps_parser.add_argument(
+        "--port", default=9000, type=int, help="Port for the DB."
+    )
+    clickhouse2deps_parser.add_argument(
```
Comment (vmarkovtsev) [inline sourced/ml/mining/__main__.py:62]: Let's hardcode it in a global variable.

File: sourced/ml/mining/__main__.py [hunk 8]
```diff
@@ -23,6 +27,53 @@ def add_parser(name, help_message):
...
+        "--table", default="uasts", help="Table name for the DB."
+    )
+    clickhouse2deps_parser.add_argument(
+        "--langs",
+        nargs="+",
+        default=CLICKHOUSE_LANGS,
+        choices=CLICKHOUSE_LANGS,
+        help="Languages to extract dependencies from.",
```
Comment (vmarkovtsev) [inline sourced/ml/mining/__main__.py]:
```suggestion
        help="Languages to consider while extracting dependencies",
```

File: sourced/ml/mining/cmd/clickhouse2deps.json [hunk 9]
```diff
@@ -0,0 +1,83 @@
+{
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.json]: I bet that YAML will be more user-friendly here.

File: sourced/ml/mining/cmd/clickhouse2deps.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+QUERY_ARGS = "clickhouse2deps.json"
+CLICKHOUSE_LANGS = [
+    "cpp",
+    "csharp",
+    "go",
+    "java",
+    "javascript",
+    "php",
+    "python",
+]  # TODO: add ruby
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.py]:
```suggestion
]  # TODO(r0mainK): add ruby
```

File: sourced/ml/mining/cmd/clickhouse2deps.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+
+def clickhouse2deps(args):
+    """
+    Extract dependencies from UASTs in a ClickhouseDB.
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.py]:
```suggestion
    Extract dependencies from UASTs in a ClickHouse DB.
```

File: sourced/ml/mining/cmd/clickhouse2deps.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    """
+    log = logging.getLogger("clickhouse2deps")
+    if not args.output_path.endswith(".asdf"):
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.py]: Let's use https://dusty.phillips.codes/2018/08/13/python-loading-pathlib-paths-with-argparse/

File: sourced/ml/mining/cmd/clickhouse2deps.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        args.output_path += ".asdf"
+    if os.path.exists(args.output_path):
+        if not args.force:
+            log.error(
+                "'%s' already exists, aborting (use force to remove).", args.output_path
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.py]:
```suggestion
                "%s already exists, aborting (use -f/--force to overwrite)", args.output_path
```

File: sourced/ml/mining/cmd/clickhouse2deps.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+            )
+            raise FileExistsError
+        log.warn("'%s' already exists, removing.", args.output_path)
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.py]:
```suggestion
        log.warn("%s already exists, overwritten", args.output_path)
```

File: sourced/ml/mining/cmd/clickhouse2deps.py [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+        os.remove(args.output_path)
+    log.info("Loading the query template ...")
+    template_loader = jinja2.FileSystemLoader(
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.py]:
```python
root = Path(__file__).parent
jinja2.FileSystemLoader(root)

(root / QUERY_ARGS).open()
```

File: sourced/ml/mining/cmd/clickhouse2deps.py [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+        user=args.user,
+        password=args.password,
+        host=args.host,
+        port=args.port,
+        database=args.database,
+    )
+    files, deps, ind_files, ind_deps = [], [], [], []
+    ind_to_langs, ind_to_repos, file_to_inds, dep_to_inds = {}, {}, {}, {}
+    for lang in args.langs:
+        log.info("Extracting the dependencies of %s files ...", lang)
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.py]:
```suggestion
        log.info("Extracting %s dependencies...", lang)
```

File: sourced/ml/mining/cmd/clickhouse2deps.py [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+        rows = client.execute_iter(
+            template.render(lang=lang, table=args.table, query_args=query_args[lang]),
+            settings={"max_block_size": args.max_block_size},
+        )
+        num_rows, num_files, num_deps = 0, 0, 0
+        for row in rows:
+            try:
+                row = [e.decode("utf-8") if not isinstance(e, str) else e for e in row]
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.py]:
```suggestion
                row = [e.decode("utf-8", errors="ignore") if not isinstance(e, str) else e for e in row]
```

thus no need to catch unicode errors

File: sourced/ml/mining/cmd/clickhouse2deps.py [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+            ind_to_langs[ind_file] = lang
+            ind_to_repos[ind_file] = repo
+            ind_files.append(ind_file)
+            if lang_dep not in dep_to_inds:
+                num_deps += 1
+                dep_to_inds[lang_dep] = len(dep_to_inds)
+                deps.append(dep)
+            ind_deps.append(dep_to_inds[lang_dep])
+        log.info(
+            "Finished extraction, retrieved %d rows, with %d distinct dependencies from %d files.",
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.py]:
```suggestion
            "Finished with %s, retrieved %d rows with %d distinct dependencies in %d files", lang,
```

File: sourced/ml/mining/cmd/clickhouse2deps.py [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+            num_rows,
+            num_deps,
+            num_files,
+        )
+    log.info(
+        "Done, retrieved %d rows, with %d distinct dependencies from %d files and %d repos.",
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.py]:
```suggestion
        "Done, retrieved %d rows with %d distinct dependencies in %d files and %d repos",
```

File: sourced/ml/mining/cmd/clickhouse2deps.py [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+        len(ind_files),
+        len(deps),
+        len(files),
+        len(set(ind_to_repos.values())),
+    )
+    log.info("Creating the sparse matrix ...")
+    matrix = coo_matrix(([True for _ in ind_files], (ind_files, ind_deps)), dtype=bool)
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.py]:
```suggestion
    matrix = coo_matrix(([True] * len(ind_files), (ind_files, ind_deps)), dtype=bool)
```

File: sourced/ml/mining/cmd/clickhouse2deps.py [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+    log.info("Creating the dependencies model ...")
+    model = Dependencies(log_level=args.log_level).construct(
+        matrix, files, deps, ind_to_langs, ind_to_repos
+    )
+    log.info("Saving dependencies model ...")
+    model.save(args.output_path, series="deps")
+    log.info("Saved model in '%s'" % args.output_path)
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.py]:
```suggestion
    log.info("Saved model to %s", args.output_path)
```

File: sourced/ml/mining/cmd/clickhouse2deps.py [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+        len(files),
+        len(set(ind_to_repos.values())),
+    )
+    log.info("Creating the sparse matrix ...")
+    matrix = coo_matrix(([True for _ in ind_files], (ind_files, ind_deps)), dtype=bool)
+    log.info("Creating the dependencies model ...")
+    model = Dependencies(log_level=args.log_level).construct(
+        matrix, files, deps, ind_to_langs, ind_to_repos
+    )
+    log.info("Saving dependencies model ...")
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.py]:
```suggestion
```

File: sourced/ml/mining/cmd/clickhouse2deps.py [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+    model.save(args.output_path, series="deps")
+    log.info("Saved model in '%s'" % args.output_path)
+    log.info("Done.")
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.py]:
```suggestion
```

File: sourced/ml/mining/models/dependencies.py [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+@register_model
+class Dependencies(Model):
+    """
+    Occurence matrix between files and dependencies.
+    """
+
+    NAME = "deps"
+    VENDOR = "source{d}"
+    DESCRIPTION = (
+        "This model contains the (binary) occurrence matrix between dependencies and files,"
```
Comment (vmarkovtsev) [inline sourced/ml/mining/models/dependencies.py]:
```suggestion
        "This model contains the (binary) cooccurrence matrix between dependencies and files,"
```

File: sourced/ml/mining/models/dependencies.py [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+        "as well as a mapping of file to languages."
```
Comment (vmarkovtsev) [inline sourced/ml/mining/models/dependencies.py]:
```suggestion
        "as well as the mapping from file names to languages."
```

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]

File: setup.py (hunk 3 above)
Comment (r0mainK) [reply setup.py]: nice !

File: sourced/ml/mining/cmd/clickhouse2deps.json (hunk 9 above)
Comment (r0mainK) [reply sourced/ml/mining/cmd/clickhouse2deps.json]: done

File: sourced/ml/mining/__main__.py (hunk 8 above)
Comment (r0mainK) [reply sourced/ml/mining/__main__.py]: Done

File: sourced/ml/mining/__main__.py (hunk 7 above)
Comment (r0mainK) [reply sourced/ml/mining/__main__.py:62]: oke

File: sourced/ml/mining/__main__.py (hunk 6 above)
Comment (r0mainK) [reply sourced/ml/mining/__main__.py]: Done

File: sourced/ml/mining/__main__.py (hunk 5 above)
Comment (r0mainK) [reply sourced/ml/mining/__main__.py]: Done

File: sourced/ml/mining/__main__.py (hunk 4 above)
Comment (r0mainK) [reply sourced/ml/mining/__main__.py]: Done

File: sourced/ml/mining/cmd/clickhouse2deps.py (hunk 10 above)
Comment (r0mainK) [reply sourced/ml/mining/cmd/clickhouse2deps.py]: Done

File: sourced/ml/mining/cmd/clickhouse2deps.py (hunk 11 above)
Comment (r0mainK) [reply sourced/ml/mining/cmd/clickhouse2deps.py]: Done

File: sourced/ml/mining/cmd/clickhouse2deps.py (hunk 13 above)
Comment (r0mainK) [reply sourced/ml/mining/cmd/clickhouse2deps.py]: Done

File: sourced/ml/mining/cmd/clickhouse2deps.py (hunk 14 above)
Comment (r0mainK) [reply sourced/ml/mining/cmd/clickhouse2deps.py]: Done

File: sourced/ml/mining/cmd/clickhouse2deps.py (hunk 16 above)
Comment (r0mainK) [reply sourced/ml/mining/cmd/clickhouse2deps.py]: Done

File: sourced/ml/mining/cmd/clickhouse2deps.py (hunk 17 above)
Comment (r0mainK) [reply sourced/ml/mining/cmd/clickhouse2deps.py]: Ah nice

File: sourced/ml/mining/cmd/clickhouse2deps.py (hunk 18 above)
Comment (r0mainK) [reply sourced/ml/mining/cmd/clickhouse2deps.py]: Done

File: sourced/ml/mining/cmd/clickhouse2deps.py (hunk 19 above)
Comment (r0mainK) [reply sourced/ml/mining/cmd/clickhouse2deps.py]: Done

File: sourced/ml/mining/cmd/clickhouse2deps.py (hunk 20 above)
Comment (r0mainK) [reply sourced/ml/mining/cmd/clickhouse2deps.py]: DOne

File: sourced/ml/mining/cmd/clickhouse2deps.py (hunk 22 above)
Comment (r0mainK) [reply sourced/ml/mining/cmd/clickhouse2deps.py]: Done

File: sourced/ml/mining/cmd/clickhouse2deps.py (hunk 21 above)
Comment (r0mainK) [reply sourced/ml/mining/cmd/clickhouse2deps.py]: Done

File: sourced/ml/mining/cmd/clickhouse2deps.py (hunk 23 above)
Comment (r0mainK) [reply sourced/ml/mining/cmd/clickhouse2deps.py]: ~Done~

File: sourced/ml/mining/cmd/clickhouse2deps.py (hunk 12 above)
Comment (r0mainK) [reply sourced/ml/mining/cmd/clickhouse2deps.py]: Done

File: setup.py [hunk 26]
```diff
@@ -1,24 +1,24 @@
...
+pkg_path = Path("sourced/ml/mining")
 sourcedml = SourceFileLoader(
-    "sourced-ml-mining", "./sourced/ml/mining/__init__.py"
+    "sourced-ml-mining", pkg_path / "__init__.py"
 ).load_module()
 
-with io.open(
-    os.path.join(os.path.dirname(__file__), "README.md"), encoding="utf-8"
-) as f:
+with io.open((Path(__file__).parent / "README.md"), encoding="utf-8") as f:
```
Comment (vmarkovtsev) [inline setup.py]:
```suggestion
with (Path(__file__).parent / "README.md").open(encoding="utf-8") as f:
```

File: sourced/ml/mining/__main__.py [hunk 27]
```diff
@@ -23,6 +28,52 @@ def add_parser(name, help_message):
...
+
+    clickhouse2deps_parser = add_parser(
+        "clickhouse2deps", "Extract dependencies from a ClickHouse DB."
+    )
+    clickhouse2deps_parser.set_defaults(handler=clickhouse2deps)
+    clickhouse2deps_parser.add_argument(
+        "-o",
+        "--output-path",
+        type=Path,
+        default=Path(__file__).absolute().parent / "deps.asdf",
```
Comment (vmarkovtsev) [inline sourced/ml/mining/__main__.py]:
```suggestion
```
(no default for the output)

**Comment by r0mainK** [comment]
@vmarkovtsev got these errors with your change to pathlib, which caused the install to fail:
- `AttributeError: 'PosixPath' object has no attribute 'rpartition'`
due to:
```
sourcedml = SourceFileLoader(
    "sourced-ml-mining", pkg_path / "__init__.py"
).load_module()
```
- after converting to string, had to the same in `package_data`:
```
error in sourced-ml-mining setup command: "values of 'package_data' dict" 
must be a list of strings (got [PosixPath('ml/mining/cmd/clickhouse2deps.yaml'), 
PosixPath('ml/mining/cmd/clickhouse2deps.sql.jinja2')])
```

due to:
```
package_data={
        "": ["LICENSE.md", "README.md"],
        "sourced": [
            p.relative_to("sourced")
            for ext in include_extensions
            for p in pkg_path.rglob("*." + ext)
        ],
    }
```

File: sourced/ml/mining/__main__.py [hunk 28]
```diff
@@ -23,6 +28,52 @@ def add_parser(name, help_message):
...
+
+    clickhouse2deps_parser = add_parser(
+        "clickhouse2deps", "Extract dependencies from a ClickHouse DB."
+    )
+    clickhouse2deps_parser.set_defaults(handler=clickhouse2deps)
+    clickhouse2deps_parser.add_argument(
+        "-o",
+        "--output-path",
+        type=Path,
+        default=Path().absolute() / "deps.asdf",
```
Comment (vmarkovtsev) [inline sourced/ml/mining/__main__.py]:
```suggestion
```
(no default value for --output)

File: sourced/ml/mining/cmd/clickhouse2deps.py [hunk 29]
```diff
@@ -0,0 +1,110 @@
...
+            raise FileExistsError
+        log.warn("%s already exists, overwritten", args.output_path)
+        os.remove(args.output_path)
+    log.info("Loading the query template ...")
+    root = Path(__file__).parent
+    template_loader = jinja2.FileSystemLoader(str(root))
+    env = jinja2.Environment(keep_trailing_newline=False)
+    template = template_loader.load(env, name=QUERY_TEMPLATE)
+    log.info("Loading the query args ...")
+    with open(root / QUERY_ARGS, "r") as fin:
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.py]:
```suggestion
    with (root / QUERY_ARGS).open() as fin:
```

File: sourced/ml/mining/cmd/clickhouse2deps.py [hunk 30]
```diff
@@ -0,0 +1,110 @@
...
+        log.info("Extracting %s dependencies...", lang)
+        rows = client.execute_iter(
+            template.render(lang=lang, table=args.table, query_args=query_args[lang]),
+            settings={"max_block_size": MAX_BLOCK_SIZE},
+        )
+        num_rows, num_files, num_deps = 0, 0, 0
+        for row in rows:
+            row = [
+                e.decode("utf-8", errors="ignore") if not isinstance(e, str) else e
+                for e in row
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.py:69]: I wonder: which are the cases when it is not a string?

File: sourced/ml/mining/cmd/clickhouse2deps.yaml [hunk 31]
```diff
@@ -0,0 +1,42 @@
+cpp:
+  filters:
+    - - "uptypes = ['uast:InlineImport']"
```
Comment (vmarkovtsev) [inline sourced/ml/mining/cmd/clickhouse2deps.yaml:3]: `- -` :thinking: ?

File: sourced/ml/mining/models/dependencies.py [hunk 32]
```diff
@@ -0,0 +1,99 @@
...
+@register_model
+class Dependencies(Model):
+    """
+    Occurence matrix between files and dependencies.
+    """
+
+    NAME = "deps"
+    VENDOR = "source{d}"
+    DESCRIPTION = (
+        "This model contains the (binary) cooccurrence matrix between dependencies and files,"
```
Comment (vmarkovtsev) [inline sourced/ml/mining/models/dependencies.py]:
```suggestion
        "Binary cooccurrence matrix between dependencies and files,"
```
(sorry; this is after the second read)

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]

File: sourced/ml/mining/models/dependencies.py (hunk 32 above)
Comment (r0mainK) [reply sourced/ml/mining/models/dependencies.py]: no problem

File: sourced/ml/mining/cmd/clickhouse2deps.py (hunk 30 above)
Comment (r0mainK) [reply sourced/ml/mining/cmd/clickhouse2deps.py:69]: it happens 4 times in the present DB, all for C++, I can track them down tonight

File: sourced/ml/mining/__main__.py (hunk 27 above)
Comment (r0mainK) [reply sourced/ml/mining/__main__.py]: Done, thought you wanted since it was in the example

File: sourced/ml/mining/cmd/clickhouse2deps.py (hunk 29 above)
Comment (r0mainK) [reply sourced/ml/mining/cmd/clickhouse2deps.py]: done

File: setup.py (hunk 26 above)
Comment (r0mainK) [reply setup.py]: done

File: sourced/ml/mining/cmd/clickhouse2deps.yaml (hunk 31 above)
Comment (r0mainK) [reply sourced/ml/mining/cmd/clickhouse2deps.yaml:3]: list of list, in the jinja we join conditions in the same list with `AND`, and then those are joind with `OR`. but I actually made a mistake

**Comment by r0mainK** [comment]
@vmarkovtsev more error due to posix, caused by modelforge this time:
```
  File "/home/romain/env/python/ml-mining/lib/python3.6/site-packages/modelforge/model.py", line 421, in _write_tree
    pos = output.tell()
AttributeError: 'PosixPath' object has no attribute 'tell'
```

It's because of this: https://github.com/src-d/modelforge/blob/97dd095d4034d20a66b561378ccbb30278a9e27e/modelforge/model.py#L413

Im gonna PR modelforge quickly, but we will have to either release a new version or change `clickhouse2deps` for it to work

**Comment by vmarkovtsev** [comment]
Yep let's hotfix Modelforge

**Comment by r0mainK** [comment]
@vmarkovtsev just updated the version for ml-core and added a Loader for yaml to get rid of:

```
YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, 
as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
```

I just checked the code, it works now :D

Also I think we may have some suppressing to do in `ml-core`, When running command-line or `make test check` I get:
```
Warning: using deprecated bblfsh v1 compatibility layer.
```

**Comment by vmarkovtsev** [comment]
bblfsh v1 is a normal warning, we should see it to remind us about our technical debt :D

File: sourced/ml/mining/models/dependencies.py [hunk 33]
```diff
@@ -0,0 +1,99 @@
...
+    register_model,
+    split_strings,
+)
+from sourced.ml.core.models.license import DEFAULT_LICENSE
+
+
+@register_model
+class Dependencies(Model):
+    """
+    Occurence matrix between files and dependencies.
```
Comment (vmarkovtsev) [inline sourced/ml/mining/models/dependencies.py:15]:
```suggestion
    Co-occurence matrix between files and dependencies.
```

**Review by vmarkovtsev — APPROVED** [review]

**Pull request merged**